### PR TITLE
[storage] Add an SQL storage driver

### LINKED ISSUE #1413: ConfigMap size limit prevents storing release
We hit a problem creating a release from a somewhat larger chart:
```
2016-10-19 09:42:22.348992 I | Updating "v1" (v1) in storage
2016-10-19 09:42:22.464534 I | configmaps: update: failed to update: ConfigMap "v1.v1" is invalid: data: Too long: must have at most 1048576 characters
2016-10-19 09:42:22.464583 I | warning: Failed to update release "v1": ConfigMap "v1.v1" is invalid: data: Too long: must have at most 1048576 characters
```

A `helm install --dry-run --debug . > manifest` produces a 600 KB file, base64 encoding that file gives a 810KB file.

It seems the serialised release somehow still gets larger then 1MB hitting the maximum size limit for a ConfigMap.

Some background: We are trying to install a complete openstack as a helm release and therefore have more then the usual handful of specs in the installed chart. The top-level chart package has a gzipped size of 113KB, unpacked its about 2MB.

I understand that this is pushing the limits of what helm is used for at the moment but in general I think the chart is not unreasonable in size.

One improvement to increase the size limit could be achieved by adding compression to the configmap storage backend. This would increase the practical size limit of release by a factor of 5 easily given that a release contains a lot of repetitive (compressible) data. Would a PR for that considered useful?

Any other ideas to fix our issue (apart from splitting up the release) ?

### LINKED ISSUE #1996: `release: not found` when I do `helm install` on a chart with some weird properties
Came across this while doing some mad science on directories that are both Helm charts and Javascript modules. [Minimal repro case is here](https://github.com/streamplace/weird-chart-bug). The end result is that Helm successfully deploys a chart and immediately forgets that it did, like so:

```
> helm install --name test .
NAME:   test
Error: getting deployed release "test": release: not found
```

It does actually create the relevant Deployment object, but it doesn't know anything about the `test` release. I have to manually go in and delete the Deployment with `kubectl`.

I'm sure the problem is related to the `node_modules` directory -- it contains a zillion files and directories for all of my Javascript dependencies. Potentially I'm exceeding some hidden limit for files allowed within a Helm chart? Obviously I should be putting `node_modules` in `.helmignore`, but presumably Helm should also be doing something more coherent upon this failure case.

### LINKED ISSUE #4215: helm not warning before adding large files to deployment and failing in inconsistent state
When the chart directory contains obvious temp or dev/test files such as the helm binary itself or binary files that are obviously larger than the allowed depl config map. helm starts deploying without a warning and then crashes deployment with the error failed to create: ConfigMap "SOME_NAME.v1" is invalid: []: Too long: must have at most 1048576 characters
this warning is not printed by the helm cli but just in the tiller logs wich makes it even harder to find.

expected behaviour: helm ignores files that are obviously senseless to it by standard. 

or: helm warns "There are large files in the chart directory, that would be bundled into your deployment. Should we add them to the .helmignore file? Y/n"

Output of `helm version`:
v2.9.1
Output of `kubectl version`:
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.7", GitCommit:"dd5e1a2978fd0b97d9b78e1564398aeea7e7fe92", GitTreeState:"clean", BuildDate:"2018-04-19T00:05:56Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.2-gke.3", GitCommit:"d2c7a2bd41036f9474287579a725dc54c904e92d", GitTreeState:"clean", BuildDate:"2018-05-23T00:19:39Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
gke

### PR DESCRIPTION
**Pull request by elafarge**
**What this PR does / why we need it**:

This commits adds the possibility to back Tiller (or the future
Tiller-less Helm CLI) with any SQL database (only postgres has been
tested so far) to store release information.

The main motivation for this commit was to use a storage backend that
would allow releases larger that 1MB in size (ConfigMap or Secret
drivers don't, because of limits on value size in the underlying etcd
key-value store).

Signed-off-by: Étienne Lafarge <etienne.lafarge@gmail.com>

Co-authored-by: Elliot Maincourt <e.maincourt@gmail.com> (@emaincourt)
Co-authored-by: Paul Borensztein <hi@0x01.fr> (@commit-master)

**Special notes for the reviewer(s)**:

* It goes without saying that we'll be glad to maintain and update this part of the codebase - in particular - to comply with the new Tiller-less Helm 3 ;-) 
* Also, we'll be glad to write an `any backend -> postgres` migration script if this backend gets promoted to GA at some point.
* Only `postgresql` has been tested so far. However, we've tried to make our code SQL dialect-agnostic (using `sqlx` and avoiding `postgres` specific SQL statements). We'll be happy to test other databases if required by the community.
* We're currently testing this feature on our pre-production environment (to provide our developers with a "staging-on-demand" feature) and will switch our production infrastructure on it if this PR gets merged

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility --> **not applicable for now, we'll be glad to write a migration script and add it to the helm CLI if needed**

closes #1413
closes #1996 
closes #4215

### CODE REVIEW

File: pkg/storage/driver/sql.go
Comment (bacongobbler) [inline pkg/storage/driver/sql.go]: Is there some way we can use the Release object similar to the other storage drivers? I understand the need for the annotations for the sql driver to work, but this means we now have [two different ways of representing Helm releases](https://github.com/helm/helm/blob/33589472cf02d8d10e96e06085d95791c0611a75/pkg/proto/hapi/release/release.pb.go#L19) in the backend, which means that any time we introduce changes to the Release object, we'll also have to update this struct as well.

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: cmd/tiller/tiller.go
Comment (bacongobbler) [inline cmd/tiller/tiller.go]: It would be nice if we validated that the dialect chosen was valid or not. For example, I could enter `postgresql` or `postgres`. Without validation we cannot inform the user which dialects tiller supports or not. As far as I can tell we only accept postgresql as a valid dialect.

File: docs/sql-storage.md
Comment (bacongobbler) [reply docs/sql-storage.md]: After looking at the code, it looks like only postgres support is baked in, as it only imports the pq driver, so I'd change this to

> The SQL backend only works with PostgreSQL at this time.

File: docs/install.md
Comment (bacongobbler) [reply docs/install.md]: Also would you mind removing the `sslmode` string? In production, one should not be running a postgres database without sslmode configured (or at least the documentation shouldn't be explicitly disabling it).

File: docs/install.md
Comment (elafarge) [reply docs/install.md]: You're totally right, an old habit of mine of pronouncing SQL as "es-queue-ale" instead of "sequel"... 

That's fixed, thanks :)

File: docs/install.md
Comment (elafarge) [reply docs/install.md]: Sure ! Both changes have been made.

File: docs/sql-storage.md
Comment (elafarge) [reply docs/sql-storage.md]: :ok_hand: . This file was actually deleted and its content left as an exercise for the reader

File: docs/sql-storage.md
Comment (elafarge) [reply docs/sql-storage.md]: :ok_hand:

File: docs/sql-storage.md
Comment (elafarge) [reply docs/sql-storage.md]: Thanks for the correction :)

File: docs/sql-storage.md
Comment (elafarge) [reply docs/sql-storage.md]: :ok_hand:

File: docs/sql-storage.md
Comment (elafarge) [reply docs/sql-storage.md]: That's totally fine to me, we just got rid of this document.

File: cmd/tiller/tiller.go
Comment (elafarge) [reply cmd/tiller/tiller.go]: Totally right, we added some extra validation on the storage driver in the `NewSQL()` method. 

As you mentioned above, since we're importing `pq` only, the current implementation is limited to postgres. That totally makes sense.

**Comment by elafarge** [comment]
Thanks a lot for the quick review @bacongobbler .

All the requested changes have been implemented so far but one.
We'll think about it over the weekend and come back to you on that matter early next week.

Have a great weekend !

File: pkg/storage/driver/sql.go
Comment (technosophos) [inline pkg/storage/driver/sql.go]: Nit: The whitespace here seems to be a mix of spaces and tabs for line prefixing.

File: pkg/storage/driver/sql.go
Comment (technosophos) [reply pkg/storage/driver/sql.go]: I'm sorta inclined to go this way, since moving the other way would involve adding writing new packages that plug into the protobuf tools, which are already the most fragile piece of Helm. So I would be okay with this... and hopefully Helm 3 will make this much easier.

File: pkg/storage/driver/sql.go
Comment (technosophos) [inline pkg/storage/driver/sql.go]: You might need to sanitize the value of `key` to avoid a Bobby Tables attack. https://www.xkcd.com/327/

**Review by technosophos — CHANGES_REQUESTED** [review]
I am really excited to see this! I have long wanted to write a driver like this, but have just never had the time/need to do it. I left a security-related comment that I think needs a change (or maybe just an explanation that the SQL driver does something I don't know about). Otherwise, I am good.

Also, thanks for setting a high precedent on how a SQL driver should be tested!

File: pkg/storage/driver/sql.go
Comment (elafarge) [reply pkg/storage/driver/sql.go]: We actually check that `key` is in this (sane) map before executing the generated SQL request: https://github.com/helm/helm/pull/5371/files#diff-9614ccdf050917eda8876d24c7cb63ceR37 at line 197. Would that be fine to you ?

File: pkg/storage/driver/sql.go
Comment (bacongobbler) [reply pkg/storage/driver/sql.go]: Sounds good to me! Thanks for the clarification. :)

**Comment by elafarge** [comment]
Thanks a lot for the code review @technosophos . We took all your suggestions into account but the last one. We indeed do not sanitize the SQL field name *but we do check that it belongs to a sane set of possible fields*: https://github.com/helm/helm/pull/5371/files#diff-9614ccdf050917eda8876d24c7cb63ceR37 . 

Regarding the precedent on testing SQL drivers, `go-sqlmock` is pretty useful to write unit tests to test the logic around SQL queries indeed :) 
However, in that case, integration tests are also pretty useful as they check that the SQL backend executes the queries as expected (executes them at all). When we first ran our contribution against a real `postgres` database, we figured out a few typos that haven't been detected by `go-sqlmock` but were obvious when running the queries against a real SQL backend :) 

Is there any plans to add such integration tests to Helm (or do they exist already and we simply didn't find them) ? If so, we'll be glad to add some :)

**Comment by callicles** [comment]
Thanks for that work! 
I recently run into that was well and I am waiting for that to be able to move to production with Helm

**Comment by elafarge** [comment]
@bacongobbler @technosophos All the requested changes have been taken into account. Let us know if there's anything else you'd like us to change / add / delete :) 

Thanks again for your insights and suggestions :)

File: pkg/storage/driver/sql.go
Comment (bacongobbler) [reply pkg/storage/driver/sql.go]: yeah, that looks fine to me!

**Review by bacongobbler — APPROVED** [review]
looks great!

**Comment by elafarge** [comment]
@technosophos @bacongobbler we just rebased our branch against `master` and fixed some conflicts in the `glide.lock` file. Let us know if there's anything else you'd like us to change :)

**Comment by callicles** [comment]
@technosophos Would be great if that could make it to the next release!

**Comment by elafarge** [comment]
(just rebased against the `master` branch, fixed a conflict in `tiller.go` and force-pushed).

**Comment by tg339** [comment]
@technosophos any update on this one :) We're excited to see this come through to help us with getting over a hurdle that's preventing us from using helm in production

**Comment by bacongobbler** [comment]
He's currently on vacation for the next week. I'd try asking in #helm-dev on Slack and ask if another maintainer would be willing to review this.

**Comment by elafarge** [comment]
(FYI, I just rebased the branch against `master`, fixed conflicts, and force-pushed)

**Comment by elafarge** [comment]
Hi @technosophos @bacongobbler ,

Could you take a look at this PR when you get the chance. All required updates have been made. We'll be glad to change/add/remove/discuss anything you feel is relevant in a timely manner so that this can make it to 2.14.0 ;-) 

Have a great day !

**Review by technosophos — APPROVED** [review]
LGTM. I'll leave it to @bacongobbler to merge when it fits the release schedule.

**Comment by bacongobbler** [comment]
I completely forgot that we made the call to bring this into the 2.14 release. I'll cut rc.2 with this PR. Thanks @elafarge!

**Comment by bacongobbler** [comment]
cherry-picked into 2.14.

**Comment by bacongobbler** [comment]
Just a heads up, but the SQL driver tests failed after merging this into master: https://circleci.com/gh/helm/helm/8204

@elafarge would you mind looking into this to fix up the breaking tests?

**Comment by bacongobbler** [comment]
Looks like it's due to the underlying gRPC layer being populated. See attached.

image

**Comment by bacongobbler** [comment]
I've pulled it out of the 2.14 release branch for the time being so we can continue to test the release candidate. @elafarge, when you have time to look into this, please reach out to me and we can see what we can do about getting this in for 2.14. Thanks!

**Comment by bacongobbler** [comment]
Found the fix; see #5706.

**Comment by elafarge** [comment]
Hi @bacongobbler . Sorry, for the late reply, I was off for the week. Glad to hear you found the fix.

Thanks a lot for the merge !! :tada:

**Comment by vicaya** [comment]
Does the sql storage driver work with helm v3?

If not, any plans to port this (and other storage drivers) to v3.

**Comment by hickeyma** [comment]
This would ned to be ported to v3. @elafarge Is this something you would be interested in doing?

**Comment by elafarge** [comment]
@hickeyma sorry, I just stumbled upon your question when cleaning up my Github notifications. I'll surely take a peek at this (maybe with @emaincourt if he's interested :angel: ) this week.

**Comment by emaincourt** [comment]
@elafarge I might actually have time to handle it. Will take a look at the v3 codebase this week 🧚‍♀️

**Comment by hickeyma** [comment]
Thanks for getting back @elafarge @emaincourt. That would be great if you can port it. Reach out if you need any help. FYI, Helm v3 branch is now `master`. Push a new PR with a reference this PR.

**Comment by emaincourt** [comment]
@hickeyma @elafarge PR here: https://github.com/helm/helm/pull/7635

**Pull request merged**
